### DISCUSSION TITLE
Support Markdown rendering

### DISCUSSION THREAD

**Discussion by khalidabuhakmeh** [spectreconsole/spectre.console#456] [Ideas]
Allow folks to take a markdown file and render it to the console.

https://github.com/charmbracelet/glow

**Comment by mcon** [comment]
I'm going to have a shot at this, looks like it's not too tricky to use https://github.com/xoofx/markdig (I'll be working here: https://github.com/mcon/spectre.console/tree/markdown - should have something of decent quality shortly)

**Comment by mcon** [comment]
PR here if anyone's interested: https://github.com/spectresystems/spectre.console/pull/204

**Comment by enginexon** [comment]
Is this still interesting topic for community? Was it already implemented?

**Comment by patriksvensson** [comment]
@enginexon A spike PR was opened at https://github.com/spectreconsole/spectre.console/pull/204, but nothing was completed. It's up for grabs.

**Comment by frankhaugen** [comment]
I like the general concept, but I think the implementation should be a "Document", that is a full width panel, and have some internals for paging etc., but leaves the content up te "extension" through a DocumentContent -type that is inheritable, or something, at least it should allow one to override the raw content and then when it hits the renderer it is in "markup", so the internet at large could then just introduce their own nuget named something like: Spectre.Console.Document.Markdown, and you will get the community to provide the extension, (Markdown might be internally made, since someone already made some code for that in above mentioned PR).

It feels like a natural addition to have some sort of "document viewer", that have in the case of markdown 1 file per page, but being able to construct a "Pagable Document Viewer" would be awesome, but it needs some actual design an forthought to make it "fit" into the whole of spectre

**Comment by FrankRay78** [comment]
Should this really be a core part of spectre.console, or would it best be located in some form of `contrib` project? (like what Cake does, here: https://github.com/cake-contrib)

I often wonder the same thing each time I consider making a spectre.console widge for [asciichart](https://github.com/NathanBaulch/asciichart-sharp).

**Reply by frankhaugen** [reply]
Conceptually I like it, but "contrib" named nugets are usually a "red flag" for many, (at least two different teams I've worked with have had bad experience with "anything goes" contrib packages), so should probably brand it "Spectre.Console.Extensions" or "Spectre.Console.Community", and then try to motivate the community to use that as a prefix.

Personally, I like granualar nugets, so I would be urked by needing to get asciichart if I just want Markdown, so a singular "contrib" project should regardless aim for granularity, (makes handling bugs easier aswell)

**Comment by crwsolutions** [comment]
Hi all, time to dust this one off? With all those AI chat responses in markup, I think this feature is more important now then it was the past. I think/ hope I have time to update the PR to the latest version. Is that something desirable?

**Reply by FrankRay78** [reply]
@crwsolutions I'm interested in this idea. Do you have specific use cases in mind?

**Reply by crwsolutions** [reply]
@FrankRay78  I am fiddling around with AI. It has a main loop to process chat responses. Like so (I do not render the stream, thats a nice to have for me):

```csharp
    while (true)
    {
        // Get user prompt and add to chat history
        AnsiConsole.Markup("[green]Your prompt: [/]");
        var userPrompt = Console.ReadLine();

        if (userPrompt == "bye")
        {
            break;
        }

        chatHistory.Add(new ChatMessage(ChatRole.User, userPrompt));

        // Stream the AI response and add to chat history
        var response = "";
        await AnsiConsole
            .Status()
            .Spinner(Spinner.Known.Dots)
            .SpinnerStyle(new Style(foreground: Color.Green))
            .StartAsync(
                "Processing...",
                async ctx =>
                {
                    await foreach (
                        var item in _chatClient.GetStreamingResponseAsync(
                            chatHistory,
                            chatOptions
                        )
                    )
                    {
                        ctx.Status("Receiving...");
                        response += item.Text;
                    }
                }
            );
        MdConsole.Render(response);
        chatHistory.Add(new ChatMessage(ChatRole.Assistant, response));
        Console.WriteLine();
    }
    AnsiConsole.MarkupLine("[red]Bye, Bye! 💖[/]");
}
```

the main markup in my case is headings, bold, lists and code blocks. The code blocks are C#, Typescript, json and SQL. My current output:

![image](https://github.com/user-attachments/assets/7fb06033-565d-4664-a3f7-d7ecfa21c8d1)

**Comment by phil-scott-78** [comment]
I gave this a go a little while ago in terms of the AI chat stuff. It gets hard quick because I think the expectation would be that the markdown would be rendered as it is streamed to the app from the LLM. Rendering incomplete markdown and then adjusting it on the fly isn't trivial, things like lists, qutoes and block code blocks all have complicated logic that needs to be accounted for as it streams, adjusting as new content and the markdown AST changes all for content that could be thousands of characters long. 

I think the amount of edge cases, moving parts and scope of this would be large enough to warrant being an entirely different project with someone with a need for this being the maintainer. It's gonna be a big chunk of code to maintain with its own set of issues. I think there is some stuff potentially Spectre has internal that might need to be made public to make it happen.

**Reply by crwsolutions** [reply]
@phil-scott-78  Streaming would be very cool, but a nice to have as far as I am concerned. I am not able to decode what they are saying in this issue where they discuss this matter, but maybe it is helpful for you? https://github.com/xoofx/markdig/issues/858

**Comment by phil-scott-78** [comment]
I dug up my old code. Probably a good starting point if someone wants to give it a go. 

```csharp

class MarkdownRenderable(string markdown) : Renderable
{
    protected override IEnumerable<Segment> Render(RenderOptions options, int maxWidth)
    {
        var ast = Markdown.Parse(markdown);
        foreach (var block in ast)
        {
            var r = RenderBlock(block).Render(options, maxWidth);
            foreach (var segment in r)
            {
                yield return segment;
            }

            yield return Segment.LineBreak;
        }
    }

    private IRenderable RenderBlock(Block block)
    {
        return block switch
        {
            ParagraphBlock paragraph => CreateTextBlock(paragraph),
            HeadingBlock heading => BuildingHeadingBlock(heading),
            ListBlock list => BuildListItemControl(list),
            QuoteBlock quoteBlock => BuildQuoteControl(quoteBlock),
            FencedCodeBlock codeBlock => BuildCodeControl(codeBlock),
            CodeBlock codeBlock => BuildCodeControl(codeBlock),
            ThematicBreakBlock => new Grid() { Width = 120 }.AddColumn().AddRow(new Rule()),
            _ => new Paragraph()
        };
    }

    private IRenderable BuildingHeadingBlock(HeadingBlock heading)
    {
        var sb = new StringBuilder();
        if (heading.Inline != null)
        {
            AddInlineContent(sb, heading.Inline);
        }

        var content = sb.ToString().EscapeMarkup();

        switch (heading.Level)
        {
            case 1:
                return new FigletText(content);
            case > 4:
                return new Text(content, Color.Yellow);
        }

        var rowChar = heading.Level == 2 ? '=' : '-';

        var headerRow = new Text(content, Color.Yellow);
        var rowRow = new Text(new string(rowChar, content.Length), Color.Yellow);

        return new Rows(headerRow, rowRow);
    }

    private IRenderable CreateTextBlock(LeafBlock textBlock)
    {
        var sb = new StringBuilder();
        if (textBlock.Inline != null)
        {
            AddInlineContent(sb, textBlock.Inline);
        }

        var grid = new Grid { Width = 120, }
            .AddColumn()
            .AddRow(new Markup(sb.ToString()));

        return grid;
    }

    private IRenderable BuildCodeControl(CodeBlock codeBlock)
    {
        var sb = new StringBuilder();
        var lines = codeBlock.Lines.Lines;

        foreach (var line in lines)
        {
            sb.AppendLine(line.ToString());
        }

        var panel = new Panel(sb.ToString().EscapeMarkup().Trim())
        {
            Border = new LeftBorder(),
            BorderStyle = Color.Blue,
            Padding = new Padding(1, 0, 0, 0),
        };

        if (codeBlock is FencedCodeBlock fencedCodeBlock && !string.IsNullOrEmpty(fencedCodeBlock.Info))
        {
            panel.Header = new PanelHeader(fencedCodeBlock.Info.EscapeMarkup());
        }

        return panel;
    }

    private IRenderable BuildQuoteControl(QuoteBlock quoteBlock)
    {
        var sb = new StringBuilder();

        foreach (var block in quoteBlock)
        {
            if (block is ParagraphBlock { Inline: not null } paragraphBlock)
            {
                AddInlineContent(sb, paragraphBlock.Inline);
            }
            else
            {
                var renderable = RenderBlock(block);
                sb.AppendLine(renderable.ToString());
            }
        }

        var panel = new Panel(sb.ToString().TrimEnd())
        {
            Border = new LeftBorder(),
            BorderStyle = new Style(Color.Green),
            Padding = new Padding(1, 1)
        };

        return panel;
    }

    private IRenderable BuildListItemControl(ListBlock listBlock)
    {
        var table = new Table().HideHeaders().Border(TableBorder.None);

        table.AddColumn(new TableColumn("Marker").Width(3));
        table.AddColumn(new TableColumn("Content"));

        var index = 1;

        foreach (var item in listBlock)
        {
            if (item is ListItemBlock listItem)
            {
                string marker = listBlock.IsOrdered ? $"{index}." : "\u25cb";
                index++;

                // Create a list of renderables for all blocks in the list item
                var contentRenderables = new List<IRenderable>();

                foreach (var block in listItem)
                {
                    // Recursively render each block and add it to the list
                    var blockRenderable = RenderBlock(block);
                    contentRenderables.Add(blockRenderable);
                }

                // Use Rows to combine all the renderables
                var contentRows = new Rows(contentRenderables);

                // Add the marker and content to the table
                table.AddRow(new Text(marker, Color.Green), contentRows);
            }
        }

        return table;
    }

    private void AddInlineContent(StringBuilder sb, ContainerInline textBlockInline)
    {
        foreach (var inline in textBlockInline)
        {
            switch (inline)
            {
                case CodeInline code:
                    sb.Append("[fuchsia]`");
                    sb.Append(code.Content.EscapeMarkup());
                    sb.Append("`[/]");
                    break;
                case LiteralInline literal:
                    sb.Append(literal.Content.ToString().EscapeMarkup());
                    break;

                case EmphasisInline emphasis:
                    switch (emphasis.DelimiterCount)
                    {
                        case 1:
                            sb.Append("[yellow]");
                            break;
                        case 2:
                            sb.Append("[blue]");
                            break;
                    }

                    AddInlineContent(sb, emphasis);
                    sb.Append("[/]");
                    break;

                case LineBreakInline:
                    sb.Append(" ");
                    break;

                default:
                    if (inline is ContainerInline containerInline)
                    {
                        AddInlineContent(sb, containerInline);
                    }

                    break;
            }
        }
    }
}

class LeftBorder : BoxBorder
{
    public override string GetPart(BoxBorderPart part)
    {
        if (part is BoxBorderPart.Left)
        {
            return "\u2502";
        }

        return " ";
    }
}
```

![image](https://github.com/user-attachments/assets/a1bc0118-d59f-4725-90ba-b4446e192212)

![image](https://github.com/user-attachments/assets/47c41934-e8d3-41ea-a0f5-1c7e99a81842)

**Reply by FrankRay78** [reply]
Remind me please - is there an allowable NuGet prefix that community extensions can use? Or is 'Spectre.Console.*' entirely reserved?

There is an open issue about creating a page on the website listing some useful extensions, and clarifying how to go about it. 

It seems this is timely to clarify.

**Reply by patriksvensson** [reply]
I think `Spectre.Console.Community` is open for all. If not, tell me and I'll file an issue with NuGet.

**Comment by crwsolutions** [comment]
Hi @FrankRay78 , I've worked a little bit more on the subject and taken the code of @mcon and @phil-scott-78 and split it. I've *reused the json superpowers library* and written an xml and csharp myself to markup these code blocks. It is still POC - ish. But you can find it here: https://github.com/crwsolutions/Spectre.Console.Extensions.Markup

It contains 4 projects:
- Spectre.Console.Xml, markup for xml, works exactly as the json one. No dependencies.
- Spectre.Console.CSharp, markup for c#, works also as the json one. Depends on Microsoft.CodeAnalysis.*
- Spectre.Console.Extensions.Markup, parses the markup. Contains the code of @mcon and @phil-scott-78
- Spectre.Console.Extensions.Markup.Example. Shows the markup, uses the example of @mcon

Is the Spectre.Console team interested in taking over this code or parts of it? I am not attached to it at all and would be happy to contribute it (and change things if you like).

![image](https://github.com/user-attachments/assets/1303c8fc-cb21-4fec-871f-4c9d31c1a928)

![image](https://github.com/user-attachments/assets/aa350f36-73e0-45aa-aa70-44539cab9a11)

![image](https://github.com/user-attachments/assets/69dbf0a0-73c6-4c8c-8b84-99fa5568f3d7)

![image](https://github.com/user-attachments/assets/d9542801-f275-46a5-8257-dae8202dfc20)

**Reply by crwsolutions** [reply]
I also added a "superpowers-" project for javascript and sql:

javascript:
![image](https://github.com/user-attachments/assets/dc5425b0-8e37-48bf-980b-7322dcce7458)

sql:
![image](https://github.com/user-attachments/assets/dcc710c8-3400-42fc-8636-88e70bb5ce01)

**Comment by crwsolutions** [comment]
I have been working on the subject for the past few days. I can imagine you don't want to have all these code renderers in your code base, because the list is endless and keeps moving. What you might consider is to manage only the markup, json and xml renderers and outsource the rest to the community.

The main part with the minimal implementation is this then:

```csharp
public sealed class MarkdownRenderable(string markdown) : Renderable
{
    public readonly Dictionary<string, Func<string, JustInTimeRenderable>> CodeblockRenderables =
        new(StringComparer.OrdinalIgnoreCase)
        {
            { "json", code => new JsonText(code) },
            { "xml", code => new XmlText(code) },
        };
```

Users can then add their own renderers to the list. 

I’m curious to hear what you think of that idea.

**Reply by SadE54** [reply]
It seems nice ! Does it support table rendering ?

**Reply by crwsolutions** [reply]
@SadE54  They were ignored, but I added it this morning:

![image](https://github.com/user-attachments/assets/9d915652-dcc1-4d9a-a8dd-b33458dc55d5)

**Reply by SadE54** [reply]
nice, Thanks ! 👌

**Comment by LittleLittleCloud** [comment]
RazorConsole has a [markdown](https://github.com/LittleLittleCloud/RazorConsole/blob/main/src/RazorConsole.Core/Components/Markdown.razor) component

Here's how it looks like from [markdown gallery sample](https://github.com/LittleLittleCloud/RazorConsole/blob/main/src/RazorConsole.Gallery/Pages/MarkdownGallery.razor)
<img width="2185" height="2119" alt="image" src="https://github.com/user-attachments/assets/d821d48b-afe2-4bcd-8d22-68d61731337a" />

**Comment by crwsolutions** [comment]
I wanted to follow up on the discussion because the topic has really stuck with me. I’m very excited about the idea of streaming character support in Spectre.Console, and while exploring it, I put together a solution that I hope can spark some discussion.

I approached it in two steps:

1. **A generic streaming tokenizer library**
   A generic library (Markdown, json, typescript etc...) for tokenizing text:
   **NuGet:** [https://www.nuget.org/packages/NTokenizers](https://www.nuget.org/packages/NTokenizers)
   **Documentation:** [https://crwsolutions.github.io/ntokenizers/](https://crwsolutions.github.io/ntokenizers/)

It has the following API:

```csharp
await MarkdownTokenizer.Create().ParseAsync(stream, onToken: async token => { /* handle tokens here */ }
```

2. **A Spectre.Console integration**
   A library that converts those tokens into live console output:
   **NuGet:** [https://www.nuget.org/packages/NTokenizers.Extensions.Spectre.Console](https://www.nuget.org/packages/NTokenizers.Extensions.Spectre.Console)
   **Documentation (WIP):** [https://crwsolutions.github.io/NTokenizers.Extensions.Spectre.Console/](https://crwsolutions.github.io/NTokenizers.Extensions.Spectre.Console/)

It has the following API:

```csharp
await AnsiConsole.Console.WriteMarkdownAsync(stream);
```

Here’s a YouTube video demonstrating the new behavior:
[https://www.youtube.com/watch?v=o4LR8MxP3rg](https://www.youtube.com/watch?v=o4LR8MxP3rg)

And here’s an AI demo showing the approach in action:
[https://www.youtube.com/watch?v=NleKdlooc_0](https://www.youtube.com/watch?v=NleKdlooc_0)

Big thanks to **@juuul**, Copilot, Qwen3, and ChatGPT for their input along the way. I’d love to hear any feedback, ideas, or thoughts from the community it would be amazing to see how this could evolve!

> _post update: Markup was renamed to Markdown in v2_

**Comment by phil-scott-78** [comment]
@crwsolutions, this is really pretty slick. nice work

**Comment by boxofyellow** [comment]
I started this a while ago (like close to when this discussion was created) and was unaware of the work folks were doing in this space.

https://github.com/boxofyellow/ConsoleMarkdownRenderer

It started as just an easy options to interject your readme into your application at run time.  But after seeing this I did some refactoring so you can now markdown inside content that you are already managing with Spectre
